### ISSUE TITLE
[Controller Request]: Atari VCS

### ISSUE THREAD

**Issue by BitterSweetcandyshop** [ndeadly/MissionControl#724]
### Controller Name

Atari VCS Wireless Moodern Controller

### Current Controller Behaviour

edit: Controller is successful in pairing with the console but buttons did not work

### Controller vendor and product ID

Atari

### Supported Features

- [x] Rumble/Vibration
- [ ] Motion Controls
- [ ] Player Number Indicator
- [x] Battery Level Reporting

### Input Report Format

All mappings are clockwise starting at noon. menu items last.
Bottom Triggers are analog apparently, I did my best to map the high and low of them

Product website, should help with menu mapping: https://atari.com/products/modern-controller
Image of button map: https://atari.com/cdn/shop/files/controller-button-layout.jpg?v=1664920637 (this shit is bright, flashbang warning)

```
[D-PAD]
DP_UP: 01 00 10 00 00 00 00 00 00 00 00 00 00 00 00 00
DP_UP_RIGHT: 01 00 20 00 00 00 00 00 00 00 00 00 00 00 00 00
DP_RIGHT:  01 00 30 00 00 00 00 00 00 00 00 00 00 00 00 00
DP_DOWN_RIGHT:  01 00 40 00 00 00 00 00 00 00 00 00 00 00 00 00
DP_DOWN: 01 00 50 00 00 00 00 00 00 00 00 00 00 00 00 00
DP_DOWN_LEFT: 01 00 60 00 00 00 00 00 00 00 00 00 00 00 00 00
DP_LEFT: 01 00 70 00 00 00 00 00 00 00 00 00 00 00 00 00
DP_UP_LEFT: 01 00 80 00 00 00 00 00 00 00 00 00 00 00 00 00

[LEFT JOYCON]
JC_PRESS: 01 40 00 00 00 00 00 00 00 00 00 00 00 00 00 00
JC_UP: 01 00 00 00 00 01 80 00 00 00 00 00 00 00 00 00
JC_RIGHT: 01 00 00 ff 7f 00 00 00 00 00 00 00 00 00 00 00
JC_DOWN: 01 00 00 00 00 ff 7f 00 00 00 00 00 00 00 00 00
JC_LEFT: 01 00 00 01 80 00 00 00 00 00 00 00 00 00 00 00

[YBAX]
Y: 01 08 00 00 00 00 00 00 00 00 00 00 00 00 00 00
B: 01 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00
A: 01 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00
X: 01 04 00 00 00 00 00 00 00 00 00 00 00 00 00 00

[RIGHT JOYCON]
JC_PRESS: 01 80 00 00 00 00 00 00 00 00 00 00 00 00 00 00
JC_UP: 01 00 00 00 00 00 00 00 00 01 80 00 00 00 00 00
JC_RIGHT: 01 00 00 00 00 00 00 ff 7f 00 00 00 00 00 00 00
JC_DOWN: 01 00 00 00 00 00 00 00 00 ff 7f 00 00 00 00 00
JC_LEFT: 01 00 00 00 00 00 00 01 80 00 00 00 00 00 00 00

[TRIGGERS] 
TOP_LEFT: 01 10 00 00 00 00 00 00 00 00 00 00 00 00 00 00
BOTTOM_LEFT_START: 01 00 00 00 00 00 00 00 00 00 00 01 00 00 00 00 
BOTTOM_LEFT_PRESSED: 01 00 00 00 00 00 00 00 00 00 00 ff 03 00 00 00

TOP_RIGHT: 01 20 00 00 00 00 00 00 00 00 00 00 00 00 00 00
BOTTOM_RIGHT_START: 01 00 00 00 00 00 00 00 00 00 00 00 00 01 00 00
BOTTOM_RIGHT_PRESSED: 01 00 00 00 00 00 00 00 00 00 00 00 00 ff 03 00

[MENU BUTTONS]
BACK_BUTTON: 01 00 01 00 00 00 00 00 00 00 00 00 00 00 00 00
ATARI_HOME: 01 00 04 00 00 00 00 00 00 00 00 00 00 00 00 00
CONTEXT_MENU: 01 00 02 00 00 00 00 00 00 00 00 00 00 00 00 00

[CHARGING] will update once I can play games to pass the time, controller lasts a good while
CHARGING (with almost full battery): 02 e3 00
CHARGE_FULL (I think): 02 59 00
POWER_SUPPLY_TOO_WEAK: 02 e4 00
after unpluging from too low power supply????: 02 09 00
```

### Additional Details

I also have the Joy-Stick, Which I can also fully map out, it has quite some odd buttons though, it's good for racing games people claim tho...

once mapped, I can play with the controller and test different charge levels if needed, my atari VCS is currently in rehab after some rough software/firmware hackingattempts.

**Comment by ndeadly** [comment]
You say the controller was not successful in pairing with the console? Did this input report format stuff come from my hid_report_viewer.nro app or elsewhere? (in the case of the former then it _did_ pair with the console). I will also need the vendor and product id of the controller that can be found from running the btdb.nro app (format xxxx:xxxx in the output)

**Comment by BitterSweetcandyshop** [comment]
> You say the controller was not successful in pairing with the console?

sorry ai had to reload the github page and re-do everything at some point. it paired perfectly fine.

**Comment by BitterSweetcandyshop** [comment]
> I will also need the vendor and product id

here ya go: 44:03:77:1a:86:6b

**Comment by ndeadly** [comment]
> > I will also need the vendor and product id
> 
> here ya go: 44:03:77:1a:86:6b
> 
> 

This is the Bluetooth address, not the vendor and product id. The one I want is in the format xxxx:xxxx

**Comment by BitterSweetcandyshop** [comment]
> This is the Bluetooth address, not the vendor and product id. The one I want is in the format xxxx:xxxx

I figured I may have had that wrong, I get off work in three hours and I'll grab it then. However I'm unsure of how to find the id.

**Comment by BitterSweetcandyshop** [comment]
just got home: 3250:1002

**Comment by ndeadly** [comment]
Try this and let me know what's working. I didn't attempt to map the battery/power info at this point because it's not so clear from what you reported what it all means. 

[MissionControl-0.9.4-atari-modern-961fa39.zip](https://github.com/ndeadly/MissionControl/files/12570697/MissionControl-0.9.4-atari-modern-961fa39.zip)

**Comment by BitterSweetcandyshop** [comment]
alright will do, just one sec

**Comment by ndeadly** [comment]
Can you make a video of the behaviour on the stick calibration screen so I can get some idea of what I've messed up?

**Comment by BitterSweetcandyshop** [comment]
> Can you make a video

sure thing, I'll upload a video later.

**Comment by BitterSweetcandyshop** [comment]
Update due to the delay: I have some small discoloration on the screen, it appears it's been spreading a little and now it's to the point it's due for a repair, so I'm waiting on parts to arrive. Turns out the lcd and digitizer where damaged.

**Comment by BitterSweetcandyshop** [comment]
it'll be up for three days, just a temporary file hoster: https://litter.catbox.moe/fys1uk.MOV

**Comment by ndeadly** [comment]
> here's the video, it'll be up for three days, just a temporary file hoster: https://litter.catbox.moe/fys1uk.MOV

What I was actually hoping to see was a recording of the analog stick calibration screen in the system menu. This shows a dot representing the current stick position, so I can visually understand what my current mapping is doing wrong. I can't really see this just navigating the system menu.

And please use a better host next time, or compress your video first before uploading. That video file was 310MB and took ~10mins to download on a fast connection.

**Comment by BitterSweetcandyshop** [comment]
> What I was actually hoping to see was a recording of the analog stick calibration screen

I made a few videos before I realized I was recording the wrong thing... I uploaded the wrong video and I'll also use a different host.

**Comment by BitterSweetcandyshop** [comment]
here we go: [discord cdn bc y not](https://cdn.discordapp.com/attachments/995898541274828872/1152043847375929374/vt0v1o731i4.mov)

**Comment by BitterSweetcandyshop** [comment]
any update?

**Comment by ndeadly** [comment]
I was dumb and forgot the scale the stick values down to the range used by the Switch. This should work.

[MissionControl-0.9.4-atari-modern-5110269.zip](https://github.com/ndeadly/MissionControl/files/12657738/MissionControl-0.9.4-atari-modern-5110269.zip)

**Comment by BitterSweetcandyshop** [comment]
works perfectly! I'll submit a issue for the atari joy-stick remake, good for racing games another time.

anyhow do you have a donation page?

**Comment by ndeadly** [comment]
For rumble, the best method I've found for figuring out the packets is to make a wireshark capture on PC (capturing USB traffic on many computers will also grab Bluetooth) and find an online gamepad tester that can test vibration and trigger that a few times during the capture. If you can find the output format via this or any other method, I can support rumble.

Likewise with the battery, if you can help figure out what the values in that second report mode mean, I'll add it. I couldn't really make sense of what you submitted before. Usually, battery values either range 0x00 - 0xff (0 - 255) or 0x00 - 0x64 (0 - 100). Maybe some of the bits are used  for charging status etc on this controller too.

> anyhow do you have a donation page?

I accept donations at https://ko-fi.com/J3J01BZZ6. I also have crypto wallets listed on the main page of my repo if you prefer.

**Comment by ndeadly** [comment]
Closing this, as support has now been included in the 0.10.0 release. You can re-open the issue or reach out to me on discord if you figure out anything regarding the battery or rumble data format.
